### PR TITLE
removed compiler internal list implementation

### PR DESCRIPTION
**Pull request by krux02**
I replaced every occurence of TLinkedList with seq, and removed the TLinkedList type entirely. It means one data structure less I need to know how it works, in order to understand the compiler code.

Most importantly: less legacy.

I hope the tests all run through, but they should I did not change anything semantically.

### CODE REVIEW

**Comment by Araq** [comment]
Fails with: nimblepkg/nimscriptsupport.nim(11, 45) Error: cannot open 'compiler/lists'

Seems like you cannot simply remove stuff other clients depend on. :-)

**Comment by krux02** [comment]
Hmm, this build should still fail. `./koch tools` fails, but that is aparently not tested

File: compiler/extccomp.nim [hunk 1]
```diff
@@ -82,7 +82,7 @@ compiler gcc:
     includeCmd: " -I",
     linkDirCmd: " -L",
     linkLibCmd: " -l$1",
-    debug: "",
+    debug: " -g ",
```
Comment (dom96) [inline compiler/extccomp.nim]: hrm, what does this flag add?

Comment (krux02) [inline compiler/extccomp.nim]: -g is a compiler flag for gcc to add debug information to the build. Without it, you can't use breakpoints in the sourcecode in debuggers like gdb. You would have to debug on the binary.

File: compiler/extccomp.nim [hunk 2]
```diff
@@ -486,11 +486,11 @@ proc resetCompilationLists* =
...
   # That's because we still need to hash check the external files
   # Maybe we can do that in checkDep on the other hand?
-  initLinkedList(externalToLink)
+  externalToLink.setLen 0
 
 proc addExternalFileToLink*(filename: string) =
-  prependStr(externalToLink, filename)
-  # BUGFIX: was ``appendStr``
+  externalToLink.insert(filename, 0)
+  #prependStr(externalToLink, filename)
```
Comment (dom96) [inline compiler/extccomp.nim]: please don't leave commented code.

File: compiler/extccomp.nim [hunk 3]
```diff
@@ -527,8 +527,8 @@ proc noAbsolutePaths: bool {.inline.} =
   # `optGenMapping` is included here for niminst.
   result = gGlobalOptions * {optGenScript, optGenMapping} != {}
 
-proc add(s: var string, many: openArray[string]) =
-  s.add many.join
+#proc add(s: var string, many: openArray[string]) =
+#  s.add many.join
```
Comment (dom96) [inline compiler/extccomp.nim]: same here https://github.com/nim-lang/Nim/pull/5371/files#r100878233

File: compiler/lists.nim [hunk 4]
```diff
@@ -7,30 +7,35 @@
 #    distribution, for details about the copyright.
 #
 
-# This module implements a generic doubled linked list.
-# TODO Remove this and replace it with something sensible
+# This module is deprecated, don't use it.
```
Comment (dom96) [inline compiler/lists.nim]: instead of deprecating this, why not just remove it?

Comment (krux02) [inline compiler/lists.nim]: Well I did, but then there were complaints from @Araq that I can't just remove stuff. When he is ok, I would be very happy, to completely remove the ``compiler/lists.nim`` file entirely. It also has a conflicting name with the lists package of the standard library.

**Comment by krux02** [comment]
@dom96 I think I addressed now all your comments, do you think that there is some reason left to not merge this pull request.

File: compiler/commands.nim [hunk 5]
```diff
@@ -335,12 +335,14 @@ proc processSwitch(switch, arg: string, pass: TCmdLinePass, info: TLineInfo;
   of "excludepath":
     expectArg(switch, arg, pass, info)
     let path = processPath(arg, info)
-    lists.excludePath(options.searchPaths, path)
-    lists.excludePath(options.lazyPaths, path)
+    
+    options.searchPaths.keepItIf( cmpPaths(it, path) != 0 )
```
Comment (dom96) [inline compiler/commands.nim]: Minor, and you don't have to fix it, but for the future don't put spaces inside the parenthesis.

**Comment by Araq** [comment]
Sorry, I changed my mind. It already causes conflicts with my faster-nimsuggest branch and I dislike the change. I didn't ask for this refactoring.

**Comment by dom96** [comment]
I disagree. Any way I can convince you to accept this?

Maintaining a separate list implementation in the compiler's sources is not easy. This PR simplifies the code significantly and most importantly makes it easier for others to contribute.

**Comment by Araq** [comment]
> Maintaining a separate list implementation in the compiler's sources is not easy.

Before it was touched, there was no maintenance whatsoever required.

Oh well, at least it removes 100 lines of code.

**Pull request merged**